File: java/christmas/src/main/java/christmas/Model/Discount.java
package christmas.Model;

import java.util.HashMap;
import java.util.Map;

public class Discount {
    private int visitDate;
    private Order order;
    private String day;
    private int DESSERT_DISCOUNT = 2023;
    private int MAIN_DISCOUNT = 2023;
    private Map<String, Integer> discounts;
    private int totalDiscount;
    private String badge;
    public Discount(Order order, int visitDate){
        this.visitDate = visitDate;
        this.order = order;
        this.day = getDayOfWeek();
        makeDiscountObject();
        calculateTotalDiscount();
        caculateBadge();
    }

    private void caculateBadge() {
        if(totalDiscount>=20000){
            badge = "산타";
        }
        else if(totalDiscount>=10000){
            badge = "트리";
        }
        else if(totalDiscount>=5000){
            badge = "별";
        }
        else{
            badge = "없음";
        }
    }
    public String getBadge(){
        return badge;
    }

    private void calculateTotalDiscount() {
        totalDiscount = discounts.values().stream().mapToInt(Integer::intValue).sum();
    }

    public int getTotalDiscount() {
        return totalDiscount;
    }

    public void makeDiscountObject() {
        discounts = new HashMap<>();
        if(order.getTotalPrice()<10000){
            discounts.put("dday", 0);
            discounts.put("weekday", 0);
            discounts.put("weekend", 0);
            discounts.put("special", 0);
            discounts.put("freebie", 0);
        }
        else{
            discounts.put("dday", calculateDdayDiscount());
            discounts.put("weekday", calculateWeekDayDiscount());
            discounts.put("weekend", calculateWeekendDiscount());
            discounts.put("special", calculateSpecialDiscount());
            discounts.put("freebie", calculateFreebieDiscount());
        }
    }

    public Map<String, Integer> getDiscountObject() {
        return discounts;
    }


    public int calculateDdayDiscount() {
        if (visitDate > 25) {
            return 0;
        }
        return 900 + visitDate * 100;
    }

    public int calculateSpecialDiscount(){
        if("일".equals(day) || visitDate==25){
            return 1000;
        }
        return 0;
    }
    public int calculateFreebieDiscount() {
        if (order.getFreebie()) {
            return 25000;
        }
        return 0;
    }

    public String getDayOfWeek() {
        String[] daysOfWeek = {"목", "금", "토", "일", "월", "화", "수"};
        return daysOfWeek[visitDate % 7];
    }

    public int calculateWeekDayDiscount() {
        if ("금".equals(day) || "토".equals(day)) {
            return 0;
        }
        return DESSERT_DISCOUNT * menuCount("디저트");
    }

    public int calculateWeekendDiscount() {
        if (!"금".equals(day) && !"토".equals(day)) {
            return 0;
        }
        return MAIN_DISCOUNT * menuCount("메인");
    }

    public int menuCount(String category) {
        return order.countMenuInCategory(category);
    }

}

File: kotlin/christmas/src/main/kotlin/christmas/model/discount.kt
package christmas.model
import christmas.model.order

class discount(private val order: order, private val visitDate: Int) {
    private val day: String = getDayOfWeek()
    private val DESSERT_DISCOUNT: Int = 2023
    private val MAIN_DISCOUNT: Int = 2023
    private var discounts: MutableMap<String, Int> = mutableMapOf()
    private var totalDiscount: Int = 0
    private var badge: String = ""

    init {
        makeDiscountObject()
        calculateTotalDiscount()
        calculateBadge()
    }

    fun getBadge(): String {
        return badge
    }

    fun calculateBadge() {
        badge = when {
            totalDiscount >= 20000 -> "산타"
            totalDiscount >= 10000 -> "트리"
            totalDiscount >= 5000 -> "별"
            else -> "없음"
        }
    }

    fun getTotalDiscount(): Int {
        return totalDiscount
    }

    private fun calculateTotalDiscount() {
        totalDiscount = discounts.values.sum()
    }

    private fun makeDiscountObject() {
        discounts = if (order.getTotalPrice() < 10000) {
            mutableMapOf(
                    "dday" to 0,
                    "weekday" to 0,
                    "weekend" to 0,
                    "special" to 0,
                    "freebie" to 0
            )
        } else {
            mutableMapOf(
                    "dday" to calculateDdayDiscount(),
                    "weekday" to calculateWeekDayDiscount(),
                    "weekend" to calculateWeekendDiscount(),
                    "special" to calculateSpecialDiscount(),
                    "freebie" to calculateFreebieDiscount()
            )
        }
    }

    fun getDiscountObject(): Map<String, Int> {
        return discounts.toMap()
    }

    fun calculateDdayDiscount(): Int {
        return if (visitDate > 25) {
            0
        } else {
            900 + visitDate * 100
        }
    }

    fun calculateSpecialDiscount(): Int {
        return if ("일" == day || visitDate == 25) {
            1000
        } else {
            0
        }
    }

    fun calculateFreebieDiscount(): Int {
        return if (order.getFreebie()) {
            25000
        } else {
            0
        }
    }

    private fun getDayOfWeek(): String {
        val daysOfWeek = arrayOf("목", "금", "토", "일", "월", "화", "수")
        return daysOfWeek[visitDate % 7]
    }

    private fun calculateWeekDayDiscount(): Int {
        return if ("금" == day || "토" == day) {
            0
        } else {
            DESSERT_DISCOUNT * menuCount("디저트")
        }
    }

    private fun calculateWeekendDiscount(): Int {
        return if ("금" != day && "토" != day) {
            0
        } else {
            MAIN_DISCOUNT * menuCount("메인")
        }
    }

    private fun menuCount(category: String): Int {
        return order.countMenuInCategory(category)
    }
}
